File: VISCompass/ObservationHistory.swift
//
//  ObservationHistory.swift
//  VISCompass
//
//  Created by Matt Clark on 06/11/2015.
//  Copyright © 2015 mattclark.net. All rights reserved.
//

import Foundation
import CoreLocation


struct Observation {
    let v: Double
    let t: NSDate
}

class ObservationHistory {
    var deltaFunc: (Double, Double) -> Double = { (v1: Double, v2: Double) -> Double in return v1 - v2 }
    let gamma = 2.0
    var window_secs: Double = 10.0
    let interval = 1.0
    var otherObservations: [Observation] = []
    var mostRecentObservation: Observation?
    
    //
    // internal interface
    //
    
    init(deltaFunc f: (Double, Double) -> Double, window_secs: Double) {
        self.deltaFunc = f
        self.window_secs = window_secs
    }
    
    func add_observation(o: Observation) {
        var new_obs = observations() + [o]
        new_obs = new_obs.sort(timesorted)
        mostRecentObservation = new_obs.removeFirst()
        otherObservations = new_obs.filter(usable)
    }
    
    func smoothed(reftime: NSDate) -> Double? {
        var iseries = interval_series(reftime)
        if iseries.count == 0 {
            return nil
        }
        var sv = iseries.removeFirst()
        for (i, v) in iseries.enumerate() {
            let delta_t = Double((i + 1)) * interval
            let delta_v = weight(delta_t) * deltaFunc(sv, v)
            sv = sv + delta_v
        }
        return sv
    }
    
    //
    // private interface
    //
    
    private func observations() -> [Observation] {
        if mostRecentObservation == nil {
            return []
        }
        var obs = [mostRecentObservation!]
        obs += otherObservations
        return obs
    }
    
    private func usable(o: Observation) -> Bool {
        let window_start = NSDate().timeIntervalSinceReferenceDate - window_secs
        return o.t.timeIntervalSinceReferenceDate > window_start
    }
    
    private func timesorted(o1: Observation, o2: Observation) -> Bool {
        return o1.t.timeIntervalSinceReferenceDate > o2.t.timeIntervalSinceReferenceDate
    }
    
    private func interval_series(reftime: NSDate) -> [Double] {
        // create a series of equally intervaled values from intermittent observations
        var s = [Double]()
        var t = reftime.timeIntervalSinceReferenceDate
        let earliest_t = t - window_secs
        for o in observations() {
            while t >= o.t.timeIntervalSinceReferenceDate && t > earliest_t {
                s.append(o.v)
                t -= interval
            }
            if t < earliest_t {
                // the time is outside the window
                break
            }
        }
        return s
    }

    private func weight(delta_t: Double) -> Double {
        if delta_t >= window_secs {
            return 0
        }
        let linear_weight = (window_secs - delta_t) / window_secs
        return pow(linear_weight, gamma)
    }
}

File: Android/app/src/main/java/com/viscompass/visompass/ObservationHistory.java
package com.viscompass.visompass;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.Enumeration;

public class ObservationHistory {
    Double gamma = 2.0;
    long window_msecs = 10000;
    long interval = 1000;
    ArrayList<Observation> otherObservations;
    Observation mostRecentObservation;

public ObservationHistory(long window_msecs){
    this.window_msecs = window_msecs;
    otherObservations = new ArrayList<Observation>();
}

public void addObservation(Observation o){
    if(mostRecentObservation != null){
        otherObservations.add(mostRecentObservation);
        otherObservations.add(o);
        Collections.sort(otherObservations);
        mostRecentObservation = otherObservations.remove(0);
    }
    else{
        mostRecentObservation = o;
    };

}

public Float smoothed(Date refDateTime){

    ArrayList<Float> iseries = interval_series(refDateTime);
    if(iseries.size() == 0){return null;};
    Float sv = iseries.remove(0);
    for(int counter = 0; counter < iseries.size();counter++){
        long delta_t = (long)(counter + 1) * interval;
        Float delta_v = weight(delta_t) * deltaFunc(sv,iseries.get(counter));
        sv = sv + delta_v;
    }
    return sv;
}

Float deltaFunc(Float v1, Float v2){
    return v2-v1;
}
ArrayList<Float> interval_series(Date refTime){
    ArrayList<Float> s = new ArrayList<Float>();
    long t = refTime.getTime();
    long earliest_t = t - window_msecs;
    s.add(mostRecentObservation.v);
    if(otherObservations != null) {
        for (int counter = 0; counter < otherObservations.size(); counter++) {
            Observation obs = otherObservations.get(counter);
            if (obs.t.getTime() < earliest_t) {
                otherObservations.remove(counter);
            } else if (t >= obs.t.getTime() && t > earliest_t) {
                s.add(obs.v);
                t -= interval;
                if (t < earliest_t) {
                    break;
                }
            }
            ;
        }
    }
return s;

}
Float weight(long delta_t){
    if(delta_t >= window_msecs){return 0.0f;};
    double linear_weight = ((double)window_msecs - (double)delta_t) / (double)window_msecs;
    return (float)Math.pow((double)linear_weight, gamma);

}



}
